The job **`test-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/7927338282/job/21643550714) [(plain)](https://github.com/rust-lang-ci/rust/commit/45665a8fc2cc0470eb8020eea7c4a8dba8fe4768/checks/21643550714/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] tests/codegen/vecdeque-drain.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/vecdeque-drain/vecdeque-drain.ll" "/checkout/tests/codegen/vecdeque-drain.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC" "--dump-input-context" "100"
--- stderr -------------------------------
--- stderr -------------------------------
/checkout/tests/codegen/vecdeque-drain.rs:14:16: error: CHECK-NEXT: expected string not found in input
// CHECK-NEXT: call void @llvm.memset
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/vecdeque-drain/vecdeque-drain.ll:190:29: note: scanning from here
 %0 = getelementptr inbounds %"alloc::collections::vec_deque::VecDeque<i32>", ptr %v, i32 0, i32 1
                            ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/vecdeque-drain/vecdeque-drain.ll:196:14: note: possible intended match here
define dso_local void @truncate(ptr noalias nocapture noundef align 4 dereferenceable(16) %v, i32 noundef %n) unnamed_addr #2 {


Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/vecdeque-drain/vecdeque-drain.ll
Check file: /checkout/tests/codegen/vecdeque-drain.rs

-dump-input=help explains the following input dump.
Input was:
<<<<<<
           .
           .
           .
           .
          90:  %11 = shl i32 %dst_pre_wrap_len.i, 2 
          91:  tail call void @llvm.memmove.p0.p0.i32(ptr nonnull align 4 %dst3.i5.i, ptr nonnull align 4 %_19.i4.i, i32 %11, i1 false), !noalias !9 
          92:  %_32.i = sub i32 %len.0, %dst_pre_wrap_len.i 
          93:  %12 = getelementptr i32, ptr %self1.i3.i, i32 %dst_pre_wrap_len.i 
          94:  %_19.i7.i = getelementptr i32, ptr %12, i32 %src.0 
          95:  %13 = shl i32 %_32.i, 2 
          96:  tail call void @llvm.memmove.p0.p0.i32(ptr nonnull align 4 %self1.i3.i, ptr nonnull align 4 %_19.i7.i, i32 %13, i1 false), !noalias !12 
          97:  br label %"_ZN5alloc11collections9vec_deque21VecDeque$LT$T$C$A$GT$9wrap_copy17h6deb735074271266E.exit" 
          98:  
          99: bb17.i: ; preds = %bb8.i 
         100:  %_35.i = sub i32 %len.0, %dst_pre_wrap_len.i 
         101:  %14 = getelementptr inbounds i8, ptr %source_deque, i32 4 
         102:  %self1.i8.i = load ptr, ptr %14, align 4, !alias.scope !2, !nonnull !1, !noundef !1 
         103:  %15 = getelementptr i32, ptr %self1.i8.i, i32 %dst_pre_wrap_len.i 
         104:  %_19.i9.i = getelementptr i32, ptr %15, i32 %src.0 
         105:  %16 = shl i32 %_35.i, 2 
         106:  tail call void @llvm.memmove.p0.p0.i32(ptr nonnull align 4 %self1.i8.i, ptr nonnull align 4 %_19.i9.i, i32 %16, i1 false), !noalias !15 
         107:  %_19.i11.i = getelementptr inbounds i32, ptr %self1.i8.i, i32 %src.0 
         108:  %dst3.i12.i = getelementptr inbounds i32, ptr %self1.i8.i, i32 %dst.0 
         109:  %17 = shl i32 %dst_pre_wrap_len.i, 2 
         110:  tail call void @llvm.memmove.p0.p0.i32(ptr nonnull align 4 %dst3.i12.i, ptr nonnull align 4 %_19.i11.i, i32 %17, i1 false), !noalias !18 
         111:  br label %"_ZN5alloc11collections9vec_deque21VecDeque$LT$T$C$A$GT$9wrap_copy17h6deb735074271266E.exit" 
         112:  
         113: bb10.i: ; preds = %bb9.i 
         114:  br i1 %dst_wraps.i, label %bb26.i, label %bb20.i 
         115:  
         116: bb11.i: ; preds = %bb9.i 
         117:  br i1 %dst_wraps.i, label %bb30.i, label %bb23.i 
         118:  
         119: bb20.i: ; preds = %bb10.i 
         120:  %18 = getelementptr inbounds i8, ptr %source_deque, i32 4 
         121:  %self1.i13.i = load ptr, ptr %18, align 4, !alias.scope !2, !nonnull !1, !noundef !1 
         122:  %_19.i14.i = getelementptr inbounds i32, ptr %self1.i13.i, i32 %src.0 
         123:  %dst3.i15.i = getelementptr inbounds i32, ptr %self1.i13.i, i32 %dst.0 
         124:  %19 = shl i32 %src_pre_wrap_len.i, 2 
         125:  tail call void @llvm.memmove.p0.p0.i32(ptr nonnull align 4 %dst3.i15.i, ptr nonnull align 4 %_19.i14.i, i32 %19, i1 false), !noalias !21 
         126:  %_40.i = sub i32 %len.0, %src_pre_wrap_len.i 
         127:  %20 = getelementptr i32, ptr %self1.i13.i, i32 %src_pre_wrap_len.i 
         128:  %dst3.i17.i = getelementptr i32, ptr %20, i32 %dst.0 
         129:  %21 = shl i32 %_40.i, 2 
         130:  tail call void @llvm.memmove.p0.p0.i32(ptr nonnull align 4 %dst3.i17.i, ptr nonnull align 4 %self1.i13.i, i32 %21, i1 false), !noalias !24 
         131:  br label %"_ZN5alloc11collections9vec_deque21VecDeque$LT$T$C$A$GT$9wrap_copy17h6deb735074271266E.exit" 
         132:  
         133: bb26.i: ; preds = %bb10.i 
         134:  %delta.i = sub i32 %dst_pre_wrap_len.i, %src_pre_wrap_len.i 
         135:  %22 = getelementptr inbounds i8, ptr %source_deque, i32 4 
         136:  %self1.i18.i = load ptr, ptr %22, align 4, !alias.scope !2, !nonnull !1, !noundef !1 
         137:  %_19.i19.i = getelementptr inbounds i32, ptr %self1.i18.i, i32 %src.0 
         138:  %dst3.i20.i = getelementptr inbounds i32, ptr %self1.i18.i, i32 %dst.0 
         139:  %23 = shl i32 %src_pre_wrap_len.i, 2 
         140:  tail call void @llvm.memmove.p0.p0.i32(ptr nonnull align 4 %dst3.i20.i, ptr nonnull align 4 %_19.i19.i, i32 %23, i1 false), !noalias !27 
         141:  %24 = getelementptr i32, ptr %self1.i18.i, i32 %src_pre_wrap_len.i 
         142:  %dst3.i22.i = getelementptr i32, ptr %24, i32 %dst.0 
         143:  %25 = shl i32 %delta.i, 2 
         144:  tail call void @llvm.memmove.p0.p0.i32(ptr nonnull align 4 %dst3.i22.i, ptr nonnull align 4 %self1.i18.i, i32 %25, i1 false), !noalias !30 
         145:  %_50.i = sub i32 %len.0, %dst_pre_wrap_len.i 
         146:  %_19.i24.i = getelementptr inbounds i32, ptr %self1.i18.i, i32 %delta.i 
         147:  %26 = shl i32 %_50.i, 2 
         148:  tail call void @llvm.memmove.p0.p0.i32(ptr nonnull align 4 %self1.i18.i, ptr nonnull align 4 %_19.i24.i, i32 %26, i1 false), !noalias !33 
         149:  br label %"_ZN5alloc11collections9vec_deque21VecDeque$LT$T$C$A$GT$9wrap_copy17h6deb735074271266E.exit" 
         150:  
         151: bb23.i: ; preds = %bb11.i 
         152:  %_43.i = sub i32 %len.0, %src_pre_wrap_len.i 
         153:  %27 = getelementptr inbounds i8, ptr %source_deque, i32 4 
         154:  %self1.i25.i = load ptr, ptr %27, align 4, !alias.scope !2, !nonnull !1, !noundef !1 
         155:  %28 = getelementptr i32, ptr %self1.i25.i, i32 %src_pre_wrap_len.i 
         156:  %dst3.i26.i = getelementptr i32, ptr %28, i32 %dst.0 
         157:  %29 = shl i32 %_43.i, 2 
         158:  tail call void @llvm.memmove.p0.p0.i32(ptr nonnull align 4 %dst3.i26.i, ptr nonnull align 4 %self1.i25.i, i32 %29, i1 false), !noalias !36 
         159:  %_19.i28.i = getelementptr inbounds i32, ptr %self1.i25.i, i32 %src.0 
         160:  %dst3.i29.i = getelementptr inbounds i32, ptr %self1.i25.i, i32 %dst.0 
         161:  %30 = shl i32 %src_pre_wrap_len.i, 2 
         162:  tail call void @llvm.memmove.p0.p0.i32(ptr nonnull align 4 %dst3.i29.i, ptr nonnull align 4 %_19.i28.i, i32 %30, i1 false), !noalias !39 
         163:  br label %"_ZN5alloc11collections9vec_deque21VecDeque$LT$T$C$A$GT$9wrap_copy17h6deb735074271266E.exit" 
         164:  
         165: bb30.i: ; preds = %bb11.i 
         166:  %delta1.i = sub i32 %src_pre_wrap_len.i, %dst_pre_wrap_len.i 
         167:  %_53.i = sub i32 %len.0, %src_pre_wrap_len.i 
         168:  %31 = getelementptr inbounds i8, ptr %source_deque, i32 4 
         169:  %self1.i30.i = load ptr, ptr %31, align 4, !alias.scope !2, !nonnull !1, !noundef !1 
         170:  %dst3.i31.i = getelementptr inbounds i32, ptr %self1.i30.i, i32 %delta1.i 
         171:  %32 = shl i32 %_53.i, 2 
         172:  tail call void @llvm.memmove.p0.p0.i32(ptr nonnull align 4 %dst3.i31.i, ptr nonnull align 4 %self1.i30.i, i32 %32, i1 false), !noalias !42 
         173:  %_55.i = sub i32 %7, %delta1.i 
         174:  %_19.i33.i = getelementptr inbounds i32, ptr %self1.i30.i, i32 %_55.i 
         175:  %33 = shl i32 %delta1.i, 2 
         176:  tail call void @llvm.memmove.p0.p0.i32(ptr nonnull align 4 %self1.i30.i, ptr nonnull align 4 %_19.i33.i, i32 %33, i1 false), !noalias !45 
         177:  %_19.i35.i = getelementptr inbounds i32, ptr %self1.i30.i, i32 %src.0 
         178:  %dst3.i36.i = getelementptr inbounds i32, ptr %self1.i30.i, i32 %dst.0 
         179:  %34 = shl i32 %dst_pre_wrap_len.i, 2 
         180:  tail call void @llvm.memmove.p0.p0.i32(ptr nonnull align 4 %dst3.i36.i, ptr nonnull align 4 %_19.i35.i, i32 %34, i1 false), !noalias !48 
         181:  br label %"_ZN5alloc11collections9vec_deque21VecDeque$LT$T$C$A$GT$9wrap_copy17h6deb735074271266E.exit" 
         182:  
         183: "_ZN5alloc11collections9vec_deque21VecDeque$LT$T$C$A$GT$9wrap_copy17h6deb735074271266E.exit": ; preds = %bb6, %bb12.i, %bb14.i, %bb17.i, %bb20.i, %bb26.i, %bb23.i, %bb30.i 
         184:  ret void 
         185: } 
         186:  
         187: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: write) 
         188: define dso_local void @clear(ptr noalias nocapture noundef writeonly align 4 dereferenceable(16) %v) unnamed_addr #1 { 
         189: "_ZN4core3ptr75drop_in_place$LT$alloc..collections..vec_deque..drain..Drain$LT$i32$GT$$GT$17h45f60a3e6730b5b5E.exit": 
         190:  %0 = getelementptr inbounds %"alloc::collections::vec_deque::VecDeque<i32>", ptr %v, i32 0, i32 1 
next:14'0                                 X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
         191:  store i64 0, ptr %0, align 4 
next:14'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         192:  ret void 
next:14'0     ~~~~~~~~~~
         193: } 
next:14'0     ~~
         194:  
next:14'0     ~
         195: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: readwrite) 
next:14'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         196: define dso_local void @truncate(ptr noalias nocapture noundef align 4 dereferenceable(16) %v, i32 noundef %n) unnamed_addr #2 { 
next:14'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
next:14'1                  ?                                                                                                                   possible intended match
         197: start: 
         198:  %0 = getelementptr inbounds %"alloc::collections::vec_deque::VecDeque<i32>", ptr %v, i32 0, i32 2 
         199:  %_4 = load i32, ptr %0, align 4, !noundef !1 
         200:  %_3 = icmp ugt i32 %_4, %n 
         201:  br i1 %_3, label %bb12.i.i.i.i, label %bb4 
         202:  
         203: bb12.i.i.i.i: ; preds = %start 
         204:  %1 = icmp eq i32 %n, 0 
         205:  br i1 %1, label %bb13.i.i.i.i, label %"_ZN4core3ptr75drop_in_place$LT$alloc..collections..vec_deque..drain..Drain$LT$i32$GT$$GT$17h45f60a3e6730b5b5E.exit" 
         206:  
         207: bb13.i.i.i.i: ; preds = %bb12.i.i.i.i 
         208:  %2 = getelementptr inbounds %"alloc::collections::vec_deque::VecDeque<i32>", ptr %v, i32 0, i32 1 
         209:  store i32 0, ptr %2, align 4, !noalias !51 
         210:  br label %"_ZN4core3ptr75drop_in_place$LT$alloc..collections..vec_deque..drain..Drain$LT$i32$GT$$GT$17h45f60a3e6730b5b5E.exit" 
         211:  
         212: "_ZN4core3ptr75drop_in_place$LT$alloc..collections..vec_deque..drain..Drain$LT$i32$GT$$GT$17h45f60a3e6730b5b5E.exit": ; preds = %bb12.i.i.i.i, %bb13.i.i.i.i 
         213:  store i32 %n, ptr %0, align 4, !noalias !51 
         214:  br label %bb4 
         215:  
         216: bb4: ; preds = %start, %"_ZN4core3ptr75drop_in_place$LT$alloc..collections..vec_deque..drain..Drain$LT$i32$GT$$GT$17h45f60a3e6730b5b5E.exit" 
         217:  ret void 
         218: } 
         219:  
         220: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: readwrite) 
         221: define dso_local void @advance(ptr noalias nocapture noundef align 4 dereferenceable(16) %v, i32 noundef %n) unnamed_addr #2 { 
Build completed unsuccessfully in 0:16:40
Build completed unsuccessfully in 0:16:40
         223:  %0 = getelementptr inbounds %"alloc::collections::vec_deque::VecDeque<i32>", ptr %v, i32 0, i32 2 
         224:  %_4 = load i32, ptr %0, align 4, !noundef !1 
         225:  %_3 = icmp ugt i32 %_4, %n 
         226:  br i1 %_3, label %bb12.i.i.i.i, label %bb3 
         227:  
         228: bb3: ; preds = %start 
         229:  %_3.i = icmp eq i32 %_4, 0 
         230:  br i1 %_3.i, label %"_ZN5alloc11collections9vec_deque21VecDeque$LT$T$C$A$GT$8truncate17hd539fcb94a8dc228E.exit", label %bb3.i.i.i 
         231:  
         232: bb3.i.i.i: ; preds = %bb3 
         233:  store i32 0, ptr %0, align 4, !alias.scope !60 
         234:  br label %"_ZN5alloc11collections9vec_deque21VecDeque$LT$T$C$A$GT$8truncate17hd539fcb94a8dc228E.exit" 
         235:  
         236: "_ZN5alloc11collections9vec_deque21VecDeque$LT$T$C$A$GT$8truncate17hd539fcb94a8dc228E.exit": ; preds = %bb3, %bb3.i.i.i 
         237:  %1 = getelementptr inbounds %"alloc::collections::vec_deque::VecDeque<i32>", ptr %v, i32 0, i32 1 
         238:  store i32 0, ptr %1, align 4 
         239:  br label %bb4 
         240:  
         241: bb12.i.i.i.i: ; preds = %start 
         242:  %new_len = sub i32 %_4, %n 
         243:  %2 = getelementptr inbounds %"alloc::collections::vec_deque::VecDeque<i32>", ptr %v, i32 0, i32 1 
         244:  %idx1.i.i.i.i.i = load i32, ptr %2, align 4, !alias.scope !63, !noalias !66, !noundef !1 
         245:  %logical_index.i.i.i.i.i = add i32 %idx1.i.i.i.i.i, %n 
         246:  %3 = load i32, ptr %v, align 4, !alias.scope !63, !noalias !66, !noundef !1 
         247:  %_9.not.i.i.i.i.i = icmp ult i32 %logical_index.i.i.i.i.i, %3 
         248:  %4 = select i1 %_9.not.i.i.i.i.i, i32 0, i32 %3 
         249:  %_0.0.i.i.i.i.i = sub i32 %logical_index.i.i.i.i.i, %4 
         250:  store i32 %_0.0.i.i.i.i.i, ptr %2, align 4, !noalias !66 
         251:  store i32 %new_len, ptr %0, align 4, !noalias !66 
         252:  br label %bb4 
         253:  
         254: bb4: ; preds = %bb12.i.i.i.i, %"_ZN5alloc11collections9vec_deque21VecDeque$LT$T$C$A$GT$8truncate17hd539fcb94a8dc228E.exit" 
         255:  ret void 
         256: } 
         257:  
         258: ; Function Attrs: nounwind 
         259: define dso_local void @remove(ptr noalias nocapture noundef align 4 dereferenceable(16) %v, i32 noundef %a, i32 noundef %b) unnamed_addr #3 { 
         260: start: 
         261:  %0 = getelementptr inbounds %"alloc::collections::vec_deque::VecDeque<i32>", ptr %v, i32 0, i32 2 
         262:  %_10 = load i32, ptr %0, align 4, !noundef !1 
         263:  %_20.i = icmp ugt i32 %a, %b 
         264:  br i1 %_20.i, label %bb11.i, label %bb12.i 
         265:  
         266: bb12.i: ; preds = %start 
         267:  %_26.i = icmp ult i32 %_10, %b 
         268:  br i1 %_26.i, label %bb13.i, label %_ZN4core5slice5index5range17hca888492a7e50df4E.exit 
         269:  
         270: bb11.i: ; preds = %start 
         271: ; call core::slice::index::slice_index_order_fail 
         272:  tail call void @_ZN4core5slice5index22slice_index_order_fail17h403db1f563ee49bbE(i32 noundef %a, i32 noundef %b, ptr noalias noundef nonnull readonly align 4 dereferenceable(16) @alloc_301554e7664bb9d070032e97fa93eaf9) #7 
         273:  unreachable 
         274:  
         275: bb13.i: ; preds = %bb12.i 
         276: ; call core::slice::index::slice_end_index_len_fail 
         277:  tail call void @_ZN4core5slice5index24slice_end_index_len_fail17hd2cb829db3c1b524E(i32 noundef %b, i32 noundef %_10, ptr noalias noundef nonnull readonly align 4 dereferenceable(16) @alloc_301554e7664bb9d070032e97fa93eaf9) #7 
         278:  unreachable 
         279:  
         280: _ZN4core5slice5index5range17hca888492a7e50df4E.exit: ; preds = %bb12.i 
         281:  %drain_len = sub i32 %b, %a 
         282:  store i32 %a, ptr %0, align 4 
         283:  %new_len = sub i32 %_10, %drain_len 
         284:  %tail_len.i.i.i.i = sub i32 %new_len, %a 
         285:  %1 = icmp eq i32 %a, 0 
         286:  %2 = icmp eq i32 %new_len, %a 
         287:  %or.cond.i.i.i.i = or i1 %1, %2 
         288:  br i1 %or.cond.i.i.i.i, label %bb12.i.i.i.i, label %bb10.i.i.i.i 
         289:  
         290: bb12.i.i.i.i: ; preds = %bb10.i.i.i.i, %_ZN4core5slice5index5range17hca888492a7e50df4E.exit 
         291:  %3 = icmp eq i32 %_10, %drain_len 
         292:  br i1 %3, label %bb13.i.i.i.i, label %bb14.i.i.i.i 
         293:  
         294: bb10.i.i.i.i: ; preds = %_ZN4core5slice5index5range17hca888492a7e50df4E.exit 
         295: ; call <<alloc::collections::vec_deque::drain::Drain<T,A> as core::ops::drop::Drop>::drop::DropGuard<T,A> as core::ops::drop::Drop>::drop::join_head_and_tail_wrapping 
         296:  tail call fastcc void @"_ZN169_$LT$$LT$alloc..collections..vec_deque..drain..Drain$LT$T$C$A$GT$$u20$as$u20$core..ops..drop..Drop$GT$..drop..DropGuard$LT$T$C$A$GT$$u20$as$u20$core..ops..drop..Drop$GT$4drop27join_head_and_tail_wrapping17he930f95a96bb2638E"(ptr noalias noundef nonnull align 4 dereferenceable(16) %v, i32 noundef %drain_len, i32 noundef %a, i32 noundef %tail_len.i.i.i.i) #8, !noalias !75 
           .
           .
>>>>>>
------------------------------------------
```

</details>

--- a/library/alloc/src/collections/vec_deque/drain.rs
+++ b/library/alloc/src/collections/vec_deque/drain.rs
@@ -121,7 +121,27 @@ impl<T, A: Allocator> Drop for Drain<'_, T, A> {
                 let head_len = source_deque.len();
                 let tail_len = new_len - head_len;
 
+                // When draining at the front (`.drain(..n)`) or at the back (`.drain(n..)`),
+                // we don't need to copy any data.
+                // Outlining this function helps LLVM to eliminate the copies in these cases.
                 if head_len != 0 && tail_len != 0 {
+                    copy_data(source_deque, drain_len, head_len, tail_len);
+                }
+
+                if new_len == 0 {
+                    source_deque.head = 0;
+                } else if head_len < tail_len {
+                    source_deque.head = source_deque.to_physical_idx(drain_len);
+                }
+                source_deque.len = new_len;
+
+                #[cold]
+                fn copy_data<T, A: Allocator>(
+                    source_deque: &mut VecDeque<T, A>,
+                    drain_len: usize,
+                    head_len: usize,
+                    tail_len: usize,
+                ) {
                     let (src, dst, len);
                     if head_len < tail_len {
                         src = source_deque.head;
@@ -137,13 +157,6 @@ impl<T, A: Allocator> Drop for Drain<'_, T, A> {
                         source_deque.wrap_copy(src, dst, len);
                     }
                 }
-
-                if new_len == 0 {
-                    source_deque.head = 0;
-                } else if head_len < tail_len {
-                    source_deque.head = source_deque.to_physical_idx(drain_len);
-                }
-                source_deque.len = new_len;
             }
         }
 

--- a/library/alloc/src/collections/vec_deque/drain.rs
+++ b/library/alloc/src/collections/vec_deque/drain.rs
@@ -109,7 +109,6 @@ impl<T, A: Allocator> Drop for Drain<'_, T, A> {
 
                 let source_deque = unsafe { self.0.deque.as_mut() };
 
-                let drain_start = source_deque.len();
                 let drain_len = self.0.drain_len;
                 let new_len = self.0.new_len;
 
@@ -119,40 +118,32 @@ impl<T, A: Allocator> Drop for Drain<'_, T, A> {
                     return;
                 }
 
-                let head_len = drain_start;
+                let head_len = source_deque.len();
                 let tail_len = new_len - head_len;
 
-                match (head_len, tail_len) {
-                    (0, 0) => {
-                        source_deque.head = 0;
-                        source_deque.len = 0;
+                if head_len != 0 && tail_len != 0 {
+                    let (src, dst, len);
+                    if head_len < tail_len {
+                        src = source_deque.head;
+                        dst = source_deque.to_physical_idx(drain_len);
+                        len = head_len;
+                    } else {
+                        src = source_deque.to_physical_idx(head_len + drain_len);
+                        dst = source_deque.to_physical_idx(head_len);
+                        len = tail_len;
+                    };
+
+                    unsafe {
+                        source_deque.wrap_copy(src, dst, len);
                     }
-                    (0, _) => {
-                        source_deque.head = source_deque.to_physical_idx(drain_len);
-                        source_deque.len = new_len;
-                    }
-                    (_, 0) => {
-                        source_deque.len = new_len;
-                    }
-                    _ => unsafe {
-                        if head_len <= tail_len {
-                            source_deque.wrap_copy(
-                                source_deque.head,
-                                source_deque.to_physical_idx(drain_len),
-                                head_len,
-                            );
-                            source_deque.head = source_deque.to_physical_idx(drain_len);
-                            source_deque.len = new_len;
-                        } else {
-                            source_deque.wrap_copy(
-                                source_deque.to_physical_idx(head_len + drain_len),
-                                source_deque.to_physical_idx(head_len),
-                                tail_len,
-                            );
-                            source_deque.len = new_len;
-                        }
-                    },
                 }
+
+                if new_len == 0 {
+                    source_deque.head = 0;
+                } else if head_len < tail_len {
+                    source_deque.head = source_deque.to_physical_idx(drain_len);
+                }
+                source_deque.len = new_len;
             }
         }
 

--- a/library/alloc/src/collections/vec_deque/drain.rs
+++ b/library/alloc/src/collections/vec_deque/drain.rs
@@ -27,8 +27,8 @@ pub struct Drain<
     drain_len: usize,
     // index into the logical array, not the physical one (always lies in [0..deque.len))
     idx: usize,
-    // number of elements after the drain range
-    tail_len: usize,
+    // number of elements remaining after dropping the drain
+    new_len: usize,
     remaining: usize,
     // Needed to make Drain covariant over T
     _marker: PhantomData<&'a T>,
@@ -41,12 +41,12 @@ impl<'a, T, A: Allocator> Drain<'a, T, A> {
         drain_len: usize,
     ) -> Self {
         let orig_len = mem::replace(&mut deque.len, drain_start);
-        let tail_len = orig_len - drain_start - drain_len;
+        let new_len = orig_len - drain_len;
         Drain {
             deque: NonNull::from(deque),
             drain_len,
             idx: drain_start,
-            tail_len,
+            new_len,
             remaining: drain_len,
             _marker: PhantomData,
         }
@@ -79,7 +79,7 @@ impl<T: fmt::Debug, A: Allocator> fmt::Debug for Drain<'_, T, A> {
         f.debug_tuple("Drain")
             .field(&self.drain_len)
             .field(&self.idx)
-            .field(&self.tail_len)
+            .field(&self.new_len)
             .field(&self.remaining)
             .finish()
     }
@@ -111,18 +111,16 @@ impl<T, A: Allocator> Drop for Drain<'_, T, A> {
 
                 let drain_start = source_deque.len();
                 let drain_len = self.0.drain_len;
-                let drain_end = drain_start + drain_len;
-
-                let orig_len = self.0.tail_len + drain_end;
+                let new_len = self.0.new_len;
 
                 if T::IS_ZST {
                     // no need to copy around any memory if T is a ZST
-                    source_deque.len = orig_len - drain_len;
+                    source_deque.len = new_len;
                     return;
                 }
 
                 let head_len = drain_start;
-                let tail_len = self.0.tail_len;
+                let tail_len = new_len - head_len;
 
                 match (head_len, tail_len) {
                     (0, 0) => {
@@ -131,10 +129,10 @@ impl<T, A: Allocator> Drop for Drain<'_, T, A> {
                     }
                     (0, _) => {
                         source_deque.head = source_deque.to_physical_idx(drain_len);
-                        source_deque.len = orig_len - drain_len;
+                        source_deque.len = new_len;
                     }
                     (_, 0) => {
-                        source_deque.len = orig_len - drain_len;
+                        source_deque.len = new_len;
                     }
                     _ => unsafe {
                         if head_len <= tail_len {
@@ -144,14 +142,14 @@ impl<T, A: Allocator> Drop for Drain<'_, T, A> {
                                 head_len,
                             );
                             source_deque.head = source_deque.to_physical_idx(drain_len);
-                            source_deque.len = orig_len - drain_len;
+                            source_deque.len = new_len;
                         } else {
                             source_deque.wrap_copy(
                                 source_deque.to_physical_idx(head_len + drain_len),
                                 source_deque.to_physical_idx(head_len),
                                 tail_len,
                             );
-                            source_deque.len = orig_len - drain_len;
+                            source_deque.len = new_len;
                         }
                     },
                 }

